### PR TITLE
Proxy asset folder to webpack development server in development mode

### PR DESCRIPTION
**Pull request by jaulz**
This change reduces the application in development mode to one external port. Note that some specific Cloud9 environment variables (i.e. process.env.C9_*) are used if available.

### CODE REVIEW

File: server/koa.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ import mount from 'koa-mount';
 import helmet from 'koa-helmet';
 import logger from 'koa-logger';
 import favicon from 'koa-favicon';
+import proxy from 'koa-proxy';
```
Comment (iam4x) [inline server/koa.js]: Move the require module into the condition, it will save time when starting server on production env.

**Comment by jaulz** [comment]
Here it is :)

File: webpack/dev.config.js [hunk 2]
```diff
@@ -6,73 +6,96 @@ import webpack from 'webpack';
...
+        'process.env': {
+          BROWSER: JSON.stringify(true),
+          NODE_ENV: JSON.stringify('development')
+        }
+      }),
+  
+      new webpack.optimize.DedupePlugin(),
+      new webpack.optimize.OccurenceOrderPlugin(),
+  
+      function () { this.plugin('done', writeStats); },
```
Comment (iam4x) [inline webpack/dev.config.js]: We can remove this line, since it's gonna use `webpackConfig.server.port` with `koa-proxy` now?

**Comment by jaulz** [comment]
No, we still need it because the app needs to know which scripts shall be loaded. Though these scripts are not created in dist but instead are in memory of the webpack dev server.

File: server/koa.js [hunk 3]
```diff
@@ -55,7 +55,14 @@ app.use(hbs.middleware({
 }));
 
 const cacheOpts = {maxAge: 86400000, gzip: true};
-app.use(mount('/assets', staticCache(path.join(__dirname, '../dist'), cacheOpts)));
+
+// Proxy asset folder to webpack development server in development mode
+if (env === 'development') {
+  var webpackConfig = require('./../webpack/dev.config');
+  app.use(mount('/assets', require('koa-proxy')({ host: `http://localhost:${webpackConfig.server.port}` })));
+} else {
```
Comment (iam4x) [inline server/koa.js]:
```
} else {
```

should be written

```
}
else {
```

to stick code style from `.jscsrc`

**Comment by iam4x** [comment]
On my mac `require('os').hostname();` is equal to: `retina.local`

You should use `localhost` directly on other then Cloud9 ENV

File: webpack/dev.config.js [hunk 4]
```diff
@@ -6,73 +6,96 @@ import webpack from 'webpack';
 import writeStats from './utils/write-stats';
 import startKoa from './utils/start-koa';
 
-const WEBPACK_HOST = 'localhost';
+const PROTOCOL = (process.env.C9_HOSTNAME) ? 'https' : 'http';
+const HOST = process.env.C9_HOSTNAME || require('os').hostname();
```
Comment (iam4x) [inline webpack/dev.config.js]: On my mac `require('os').hostname();` is equal to: `retina.local`

You should use `localhost` directly on other then Cloud9 ENV

**Comment by jaulz** [comment]
Thanks for the hints :)

File: webpack/dev.config.js [hunk 5]
```diff
@@ -6,73 +6,96 @@ import webpack from 'webpack';
...
-  module: {
-    preLoaders: [
-      {
-        test: /\.js$|.jsx$/,
-        exclude: /node_modules/,
-        loaders: ['eslint', 'jscs']
+  server: {
+    port: WEBPACK_PORT,
+    options: {
+      publicPath: '/',
```
Comment (iam4x) [inline webpack/dev.config.js]: On my mac I have to edit:

```
publicPath: PUBLIC_PATH
```

File: webpack/dev.config.js [hunk 6]
```diff
@@ -6,73 +6,96 @@ import webpack from 'webpack';
 import writeStats from './utils/write-stats';
 import startKoa from './utils/start-koa';
 
-const WEBPACK_HOST = 'localhost';
+const PROTOCOL = (process.env.C9_HOSTNAME) ? 'https' : 'http';
+const HOST = process.env.C9_HOSTNAME || 'localhost';
+const PORT = (process.env.C9_HOSTNAME) ? '443' : '80';
```
Comment (iam4x) [inline webpack/dev.config.js]: Port should be `3001` on local

**Comment by iam4x** [comment]
After the lasts changes I will merge it. :+1:

**Comment by jaulz** [comment]
Can you please check again? In the future it might even make sense to have a better configuration management. One central place for all configurations for different environments.

**Comment by jaulz** [comment]
Please wait with the merge as I have to check tomorrow again. I'm facing an error in the build process:
```
Error: Deprecated option experimental: use `--stage 0`/`{ stage: 0 }`

**Comment by jaulz** [comment]
Any idea what could be wrong? I removed all babel modules again and reinstalled them though I still get the following error:
```
ERROR in Loader /home/ubuntu/workspace/node_modules/babel/index.js?experimental didn't return a function
 @ multi app

**Comment by jaulz** [comment]
It seems we were working on the same file and I saw that you removed the experimental flag. Now it does not return any error for me as well and it is working on my side.

**Comment by iam4x** [comment]
@jaulz I pull the branch, check and it will be good for merging! Thank's!

File: webpack/dev.config.js [hunk 7]
```diff
@@ -6,73 +6,96 @@ import webpack from 'webpack';
...
+          loaders: ['react-hot', 'babel']
+        },
+        {
+          test: /\.scss$/,
+          loader: 'style!css?sourceMap!autoprefixer?browsers=last 2 version!sass?outputStyle=expanded&sourceMap'
+        }
+      ]
+    },
+    plugins: [
+
```
Comment (iam4x) [inline webpack/dev.config.js]: Remove the whitespaces line 77, 81, 88, 91 and 94

File: webpack/dev.config.js [hunk 8]
```diff
@@ -6,73 +6,96 @@ import webpack from 'webpack';
...
+      function () { this.plugin('done', writeStats); },
+      function () { this.plugin('done', startKoa); }
+  
+    ],
+    resolve: {
+      extensions: ['', '.js', '.json', '.jsx'],
+      modulesDirectories: ['node_modules', 'app']
+    }
   }
 }
```
Comment (iam4x) [inline webpack/dev.config.js]: add `;` at the end of the object

Comment (iam4x) [inline webpack/dev.config.js]: add `;` to the end of the object

**Comment by iam4x** [comment]
After theses last two things I'm :ok_hand:

**Comment by jaulz** [comment]
Done, thanks :) 
Shouldn't these points pop up by the preloaders?

**Comment by iam4x** [comment]
@jaulz webpack don't compile the code for configuration && server, haven't found a clean way of doing it yet. It's in the roadmap :)

**Pull request merged**